File: python/1620. Coordinate With Maximum Network Quality.py
class Solution(object):
    def bestCoordinate(self, towers, radius):
        """
        :type towers: List[List[int]]
        :type radius: int
        :rtype: List[int]
        """
        def dis(x1, y1, x2, y2):
            return math.sqrt((x1-x2)**2 + (y1-y2)**2)

        max_val = float('-inf')
        ans = None
        N = 51
        for i in range(N):
            for j in range(N):
                total = 0
                for x, y, q in towers:
                    d = dis(i, j, x, y)
                    if d > radius: continue
                    total += int(q / (1 + d))
                if total > max_val:
                    max_val = total
                    ans = [i, j]
        return ans

File: go/src/leetcode/1620/1620. Coordinate With Maximum Network Quality.go
/**
@author ZhengHao Lou
Date    2022/11/02
*/
package main

import (
	"fmt"
	"math"
)

/**
https://leetcode.cn/problems/coordinate-with-maximum-network-quality/
*/
const N = 50

func bestCoordinate(towers [][]int, radius int) (ans []int) {
	g := make([][]int, N)
	for i := range g {
		g[i] = make([]int, N)
	}

	for i := 0; i <= N; i++ {
		for j := 0; j <= N; j++ {
			for _, tower := range towers {
				d2 := (i-tower[0])*(i-tower[0]) + (j-tower[1])*(j-tower[1])
				if d2 <= radius*radius {
					g[i][j] += int(float64(tower[2]) / float64((1 + math.Sqrt(float64(d2)))))
				}
			}
		}
	}

	ans = []int{0, 0}
	var mx int
	for i := range g {
		for j := range g[i] {
			if g[i][j] > mx {
				mx = g[i][j]
				ans = []int{i, j}
			}
		}
	}
	fmt.Println(ans)
	return ans
}

func main() {
	bestCoordinate([][]int{{1, 2, 5}, {2, 1, 7}, {3, 1, 9}}, 2)
	bestCoordinate([][]int{{23, 11, 21}}, 9)
	bestCoordinate([][]int{{1, 2, 13}, {2, 1, 7}, {0, 1, 9}}, 2)
}
